notebook/static/notebook/js/commandpalette.js

```diff
@@ -0,0 +1,168 @@
+
+define(function(require){
+    "use strict";
+
+    var QH = require("notebook/js/quickhelp");
+    var $ = require("jquery");
+
+    /**
+     * Humanize the action name to be consumed by user.
+     * internaly the actions anem are of the form
+     * <namespace>.<description-with-dashes>
+     * we drop <namesapce> and replace dashes for space.
+     */
+    var humanize_action_id = function(str) {
+      return str.split('.')[1].replace(/-/g, ' ').replace(/_/g, '-');
+    };
+
+    /**
+     * given an action id return 'command-shortcut', 'edit-shortcut' or 'no-shortcut'
+     * for the action. This allows us to tag UI in order to visually distinguish
+     * wether an action have a keybinding or not.
+     **/
+    var get_mode_for_action_id = function(name, notebook) {
+      var shortcut = notebook.keyboard_manager.command_shortcuts.get_shortcut_for_action_name(name);
+      if (shortcut) {
+        return 'command-shortcut';
+      }
+      shortcut = notebook.keyboard_manager.edit_shortcuts.get_shortcut_for_action_name(name);
+      if (shortcut) {
+        return 'edit-shortcut';
+      }
+      return 'no-shortcut';
+    };
+
+    var CommandPalette = function(notebook) {
+        if(!notebook){
+          throw new Error("CommandPalette take a notebook non-null mandatory arguement");
+        }
+
+        var form = $('<form/>').css('background', 'white');
+        var container = $('<div/>').addClass('typeahead-container');
+        var field = $('<div/>').addClass('typeahead-field');
+        var input = $('<input/>').attr('type', 'search').css('outline', 'none');
+
+        field
+          .append(
+            $('<span>').addClass('typeahead-query').append(
+              input
+            )
+          )
+          .append(
+            $('<span/>').addClass('typeahead-button').append(
+              $('<button/>').attr('type', 'submit').append(
+                $('<span/>').addClass('typeahead-search-icon')
+              )
+            )
+          );
+
+        container.append(field);
+        form.append(container);
+
+        var before_close = function() {
+          // little trick to trigger early in onsubmit
+          // when the action called pop-up a dialog
+          // insure this function is only called once
+          if (before_close.ok) {
+            return;
+          }
+          var cell = notebook.get_selected_cell();
+          if (cell) {
+            cell.select();
+          }
+          if (notebook.keyboard_manager) {
+            notebook.keyboard_manager.enable();
+            notebook.keyboard_manager.command_mode();
+          }
+          before_close.ok = true; // avoid double call.
+        };
+
+        var mod = $('<div/>').addClass('modal').append(
+          $('<div/>').addClass('modal-dialog')
+          .append(
+            $('<div/>').addClass('modal-content').append(
+              $('<div/>').addClass('modal-body')
+              .css('padding', '7px')
```

Comment:
You use .css in a few places, you should give the outer most div a `command-palette` class and style through less.  That will make it easier for users to add custom styling too.